### DISCUSSION TITLE
What are the recommended parameters to work with Firefox's built-in AI commands?

### DISCUSSION THREAD

**Discussion by TFWol** [LostRuins/koboldcpp#1256] [Q&A] [answered]
Since the implementation of Firefox's built-in AI [sort of sucks](https://github.com/Mozilla-Ocho/llamafile/issues/640#issuecomment-2499933928), I tried pointing it at koboldcpp instead of Mozilla's lamafile format.
Doing that fails to send any text when using Firefox's commands, like Summarize.

What parameters should I run koboldcpp with for it to be compatible?

**Comment by jabberjabberjabber** [comment]
It is sending the text via url ("?q=text") so it isn't going to work with Koboldcpp.

**Reply by TFWol** [reply]
Darn, I don't know why firefox has to try and be different in the worst way.

**Reply by LostRuins** [reply]
Sorry, what seems to be the issue? Right now koboldcpp does support URL query params. Are you saying it conflicts with Firefox?

Try this link
http://localhost:5001/?query=who+is+bill+gates

**Reply by TFWol** [reply]
> Sorry, what seems to be the issue? Right now koboldcpp does support URL query params. Are you saying it conflicts with Firefox?
> 
> Try this link http://localhost:5001/?query=who+is+bill+gates

I'm glad you mentioned this; it's so close to working!

The problem is firefox's built-in method makes queries like this:

`http://localhost:5001/?q=I’m+on+page+“Test+Webpage”+with+“foobar”+selected.%0A%0APlease+summarize+the+selection+using+precise+and+concise+language.+Use+headers+and+bulleted+lists+in+the+summary%2C+to+make+it+scannable.+Maintain+the+meaning+and+factual+accuracy.`

Where it uses `?q=` vs koboldcpp's `?query=`

Firefox doesn't let me modify that anywhere I could find, but I did try manually changing it in the url bar just now to:

`http://localhost:5001/?query=I’m+on+page+“Test+Webpage”+with+“foobar”+selected.%0A%0APlease+summarize+the+selection+using+precise+and+concise+language.+Use+headers+and+bulleted+lists+in+the+summary%2C+to+make+it+scannable.+Maintain+the+meaning+and+factual+accuracy.`

Which does work.

Is there a setting in kcpp that let's me specify the query parameter?

That is, without needing to modify the source code every time the program is updated?

**Reply by LostRuins** [reply]
Right now `query` is hardcoded. It's annoying that firefox only uses `q`... are you sure that cannot be changed in firefox? 

I'm worried that if I add `q` to koboldcpp as well, it would conflict with other browsers that may use `q` for something else.

**Reply by TFWol** [reply]
> It's annoying that firefox only uses q... are you sure that cannot be changed in firefox?

Nothing that I could easily find, even when I started going through the source code [here](https://searchfox.org/mozilla-central/source/toolkit/components/ml) and [here](https://searchfox.org/mozilla-central/source/browser/components/genai).
The only mention of the parameter for AI operations is [at this line](https://searchfox.org/mozilla-central/source/browser/components/genai/GenAI.sys.mjs#774) where it mentions a request header, but it's very confusing.

> I'm worried that if I add q to koboldcpp as well, it would conflict with other browsers that may use q for something else.

I was thinking you could add like a commandline option or something that lets you type the desired parameter to use.

**Reply by TFWol** [reply]
I tried a hacky workaround via userscript that automatically replaces the query parameter, but the only way to get it functional is enabling **Autosave Session**.
This makes it keep prompting me with an annoying `You already have an existing persistent story. Do you want to overwrite it?` (can I turn this prompt off?) which I will need to also automate "clicking"

If **Autosave Session** is not on and the userscript runs, it just doesn't work and errors out with:
`Error while submitting prompt: TypeError: NetworkError when attempting to fetch resource.`

**Reply by LostRuins** [reply, marked as answer]
okay i'll add `q` as an alias for `query`

**Comment by LostRuins** [comment]
Now available in the latest release!

**Reply by TFWol** [reply]
Thanks! I've been testing it out.
The core part works so it loads the right url with parameters.

The only issue I see is if I turn off **Autosave session.**
It will break and always give the error 
```
Error Encountered
Error while submitting prompt: TypeError: NetworkError when attempting to fetch resource.
```

![image](https://github.com/user-attachments/assets/fb497142-b504-4c1a-8afa-af9224bc96f9)

Looking at the web traffic, it looks like it tries to connect to **aihorde.net**, which it should not be because I have that disabled and blocked.

![image](https://github.com/user-attachments/assets/825bd69b-176a-413a-a978-f6aa6b785719)

This connection attempt does not occur if **Autosave session** is enabled.
![image](https://github.com/user-attachments/assets/8698176e-ce26-4284-8131-280c326efb8a)

![image](https://github.com/user-attachments/assets/1572edf8-3ef9-4e5a-be99-deae95d09045)

**Comment by LostRuins** [comment]
Should be fixed, please update.

**Reply by TFWol** [reply]
That got rid of the external traffic attempts and error encountered message, but the other issues are pretty much the same; if **Autosave session** is checked, it still works, but then it'll ask me `You already have an existing persistent story. Do you want to overwrite it?` popup every time I want to summarize text.

If unchecked, there is no indication of anything happening except in the command prompt window and firefox's dev console. Before it would at least have the submitted text to process.
![image](https://github.com/user-attachments/assets/539c60ea-b43a-4ff9-a2cf-d59b7e2f700c)
![image](https://github.com/user-attachments/assets/ccacb848-771d-4b7d-b1ac-9dec468b0e74)

I was able to reproduce the problem on a couple of test machines. No idea about this one. Odd firefox glitch maybe?

Since enabling Autosave at least makes it function, I might just keep it on.
If that's the case, could you please include an option to disable this message so I'm not spammed?:

![image](https://github.com/user-attachments/assets/f6b8dcbd-6341-4154-8c58-c00dfd1f5fe3)

I'd be content with that 😃

**Reply by TFWol** [reply]
Maybe another localsettings to compare on line 5138?:

https://github.com/LostRuins/koboldcpp/blob/5f8f483faec04ea2a89429e151b17403eacb2d9c/klite.embd#L5138-L5144
